File: src/main/java/de/tsenger/vdstools/idb/IcaoBarcode.java
package de.tsenger.vdstools.idb;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.security.cert.CertificateException;

import org.bouncycastle.util.encoders.Base32Encoder;

import de.tsenger.vdstools.DataEncoder;
import de.tsenger.vdstools.DataParser;

public class IcaoBarcode {
	public static final String BARCODE_IDENTIFIER = "NDB1";

	char barcodeFlag = 0x41;
	IdbPayload barcodePayload;

	public IcaoBarcode(char barcodeFlag, IdbPayload barcodePayload) {
		this.barcodeFlag = barcodeFlag;
		this.barcodePayload = barcodePayload;
	}

	public IcaoBarcode(boolean isSigned, boolean isZipped, IdbPayload barcodePayload) {
		if (isSigned)
			barcodeFlag = (char) (barcodeFlag + 0x01);
		if (isZipped)
			barcodeFlag = (char) (barcodeFlag + 0x02);
		this.barcodePayload = barcodePayload;
	}

	public boolean isSigned() {
		return ((byte) (((byte) barcodeFlag) - 0x41) & 0x01) == 0x01;
	}

	public boolean isZipped() {
		return ((byte) (((byte) barcodeFlag) - 0x41) & 0x02) == 0x02;
	}

	public IdbPayload getPayLoad() {
		return barcodePayload;
	}

	public String getEncoded() throws IOException {
		StringBuffer strBuffer = new StringBuffer(BARCODE_IDENTIFIER);
		strBuffer.append(barcodeFlag);

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		if (isZipped()) {
			bos.write(DataEncoder.zip(barcodePayload.getEncoded()));
		} else {
			bos.write(barcodePayload.getEncoded());
		}
		byte[] payloadBytes = bos.toByteArray();
		bos.reset();

		Base32Encoder base32 = new Base32Encoder();
		base32.encode(payloadBytes, 0, payloadBytes.length, bos);
		String base32EncodedPayload = bos.toString().replace("=", "");

		strBuffer.append(base32EncodedPayload);
		return strBuffer.toString();
	}

	public static IcaoBarcode fromString(String barcodeString) throws IOException, CertificateException {
		StringBuffer strBuffer = new StringBuffer(barcodeString);
		if (!strBuffer.substring(0, 4).matches(BARCODE_IDENTIFIER))
			throw new IllegalArgumentException("Didn't found an ICAO Barcode in the given String: " + barcodeString);
		char barcodeFlag = strBuffer.charAt(4);
		boolean isSigned = ((byte) (((byte) barcodeFlag) - 0x41) & 0x01) == 0x01;
		boolean isZipped = ((byte) (((byte) barcodeFlag) - 0x41) & 0x02) == 0x02;

		StringBuffer base32EncodedPayload = new StringBuffer(strBuffer.substring(5));
		while (base32EncodedPayload.length() % 8 != 0) {
			base32EncodedPayload.append("=");
		}

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		Base32Encoder base32 = new Base32Encoder();
		base32.decode(base32EncodedPayload.toString(), bos);
		byte[] payloadBytes = bos.toByteArray();

		if (isZipped) {
			payloadBytes = DataParser.unzip(payloadBytes);
		}

		IdbPayload payload = IdbPayload.fromByteArray(payloadBytes, isSigned);
		return new IcaoBarcode(barcodeFlag, payload);

	}
}

File: src/commonMain/kotlin/de/tsenger/vdstools_mp/idb/IcaoBarcode.kt
package de.tsenger.vdstools_mp.idb


import de.tsenger.vdstools_mp.Base32
import de.tsenger.vdstools_mp.DataEncoder
import de.tsenger.vdstools_mp.DataParser

class IcaoBarcode {
    private var barcodeFlag: Char = 0x41.toChar()
    var payLoad: IdbPayload

    constructor(barcodeFlag: Char, barcodePayload: IdbPayload) {
        this.barcodeFlag = barcodeFlag
        this.payLoad = barcodePayload
    }

    constructor(isSigned: Boolean, isZipped: Boolean, barcodePayload: IdbPayload) {
        if (isSigned) barcodeFlag = (barcodeFlag.code + 0x01).toChar()
        if (isZipped) barcodeFlag = (barcodeFlag.code + 0x02).toChar()
        this.payLoad = barcodePayload
    }

    val isSigned: Boolean
        get() = (((barcodeFlag.code.toByte()) - 0x41).toByte().toInt() and 0x01) == 0x01

    val isZipped: Boolean
        get() = (((barcodeFlag.code.toByte()) - 0x41).toByte().toInt() and 0x02) == 0x02

    val encoded: String
        get() {
            val strBuffer = StringBuilder(BARCODE_IDENTIFIER)
            strBuffer.append(barcodeFlag)

            val payloadBytes: ByteArray = if (isZipped) {
                DataEncoder.zip(payLoad.encoded)
            } else {
                payLoad.encoded
            }

            val base32EncodedPayload = Base32.encode(payloadBytes).replace("=", "")

            strBuffer.append(base32EncodedPayload)
            return strBuffer.toString()
        }

    companion object {
        const val BARCODE_IDENTIFIER: String = "NDB1"

        fun fromString(barcodeString: String): Result<IcaoBarcode> {
            val strBuffer = StringBuilder(barcodeString)

            if (!strBuffer.substring(0, 4).matches(BARCODE_IDENTIFIER.toRegex())) {
                return Result.failure(IllegalArgumentException("Didn't found an ICAO Barcode in the given String: $barcodeString"))
            }

            val barcodeFlag = strBuffer[4]
            val isSigned = (((barcodeFlag.code.toByte()) - 0x41).toByte().toInt() and 0x01) == 0x01
            val isZipped = (((barcodeFlag.code.toByte()) - 0x41).toByte().toInt() and 0x02) == 0x02

            val base32EncodedPayload = StringBuilder(strBuffer.substring(5))
            while (base32EncodedPayload.length % 8 != 0) {
                base32EncodedPayload.append("=")
            }

            var payloadBytes = Base32.decode(base32EncodedPayload.toString())

            if (isZipped) {
                payloadBytes = DataParser.unzip(payloadBytes)
            }

            val payload = IdbPayload.fromByteArray(payloadBytes, isSigned)
            return Result.success(IcaoBarcode(barcodeFlag, payload))
        }
    }
}
